Add label to text input fields on form

brezocordero opened the pull request:
This PR adds `<label>` to text input fields in the forms for email, last name, and first name, in the front end.

If the text form has a placeholder, the label is not visible using the class `.screen-reader-text`
The forms also have a unique ID to use on the `for` attribute on the label.

To test, create a subscription form that is displayed on the front end.
Inspect the form and check that:
- The text inputs have a unique id
- Each text input has a label with a `for` attribute using the input id
- Labels on text input fields with placeholders use the class `.screen-reader-text`

[MAILPOET-4312]

[MAILPOET-4312]: https://mailpoet.atlassian.net/browse/MAILPOET-4312?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

brezocordero committed:
Add label to text form

[MAILPOET-3331]

triple0t commented on mailpoet/lib/Form/Block/Text.php:
```diff
@@ -34,27 +34,34 @@ public function __construct(
       $autocomplete = 'email';
     }
 
     if (in_array($block['id'], ['email', 'last_name', 'first_name'], true)) {
+
       $automationId = 'data-automation-id="form_' . $this->wp->escAttr($block['id']) . '" ';
+
+      if (isset($formSettings['id'])) {
+        $id = 'id="form_' . $this->wp->escAttr($block['id']) . '_' . $this->wp->escAttr($formSettings['id']) . '" ';
+      }
     }
```
I'm curious why we are restricting **id** and **htmlFor** to only this array ['email', 'last_name', 'first_name'].
What about custom field text input types?

brezocordero replied:
@triple0t Good question!
I am not sure of why the restriction is there. It is coming from this [commit](https://github.com/mailpoet/mailpoet/pull/2634/commits/bbfd137a3c2e514df66c774a815981dfbdf2d69b). 
There is another issue in the epic that targets form controls more widely, I have added a note about custom text fields there and specified better the scope of this issue.

triple0t replied:
Hi @brezocordero 
I mean this was added specifically in this commit https://github.com/mailpoet/mailpoet/pull/4171/commits/09ac83cf8efda93997a9ae16f44daf337aa8cdf5#diff-53657c1788edfbe586720f9a5024fd9578cac10853dd2c4602daea4dd2bcdb7aR48-R50
and I was wondering why you scoped the ID to work only under the array 
```
 if (isset($formSettings['id'])) {
        $id = 'id="form_' . $this->wp->escAttr($block['id']) . '_' . $this->wp->escAttr($formSettings['id']) . '" ';
  }
```

I can see [another ticket](https://mailpoet.atlassian.net/browse/MAILPOET-3729) in the backlog addressing this issue so I guess it's fine.

BTW, I can see the same errors report when I test with https://achecker.achecks.ca/checker/index.php (Mostly for forms with custom fields)
<img width="1369" alt="Screenshot 2022-06-03 at 9 49 19 PM" src="https://user-images.githubusercontent.com/30554163/171950939-74f97481-1dc9-4a65-962f-77baebb626ec.png">
<img width="1407" alt="Screenshot 2022-06-03 at 9 49 27 PM" src="https://user-images.githubusercontent.com/30554163/171950964-b5ebefd9-a66f-4de3-b1b6-39f6acbc3b0f.png">

I will add these notes to the ticket and move this PR to QA.

Thank you

brezocordero replied:
@triple0t I made the changes and updated the other ticket. Your point made more sense.

triple0t replied:
@brezocordero Thank you for working on it 👏🏾

triple0t commented on mailpoet/lib/Form/Block/BlockRendererHelper.php:
```diff
@@ -104,28 +104,39 @@ public function getInputValidation(array $block, array $extraRules = [], ?int $f
-    }
+
     $automationId = null;
     if (in_array($block['id'], ['email', 'last_name', 'first_name'], true)) {
       $automationId = 'data-automation-id="form_' . $block['id'] . '_label" ';
+      if (isset($formSettings['id'])) {
+        $forId = 'for="form_' . $block['id'] . '_' . $formSettings['id'] . '" ';
+      }
     }
```
The same item here.

brezocordero replied:
Answered above.

triple0t reviewed (commented):
Looks good @brezocordero  

Just a note.

Kindly find below

brezocordero committed:
Add label and id also to custom fields

[MAILPOET-3331]

triple0t committed:
Fix screen-reader-text class does not exist in iFrame

screen-reader-text CSS class only exists within the WordPress environment

the class does not exist when using iFrame forms due to these being used outside the WordPress context

prefixing with mailpoet-* to not interfere with the default WordPress screen-reader-text class

MAILPOET-4312

triple0t commented:
I discovered an issue with the class `.screen-reader-text`
The CSS class only exists within the WordPress environment and due to that, iFrame form labels were showing when they should not be visible.
I fixed this in https://github.com/mailpoet/mailpoet/pull/4171/commits/fc3345a50f88ffcb9b618efdef663934d0049eea

<img width="874" alt="Screenshot 2022-06-06 at 9 40 34 AM" src="https://user-images.githubusercontent.com/30554163/172146554-73c1998d-ed5f-4a82-ae43-e48cfd800f36.png">

rohankacharegs commented:
GlobalStep Verified
